File: packages/apple/Example/LocanaraExample/components/pages/FrameworkShowcase/GuardrailDemo.swift
import SwiftUI
import Locanara

/// Demonstrates Guardrails — wrapping chains with input validation and content safety
struct GuardrailDemo: View {
    @EnvironmentObject var appState: AppState
    @State private var inputText = "Apple Intelligence is a personal intelligence system that puts powerful generative models at the core of iPhone, iPad, and Mac."
    @State private var maxCharacters = 500
    @State private var blockedPatternsText = "password, SSN, credit card"
    @State private var resultText: String?
    @State private var blockedReason: String?
    @State private var isLoading = false
    @State private var errorMessage: String?

    private var isAIAvailable: Bool {
        appState.currentEngine != .none && appState.isModelReady
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 20) {
                if !isAIAvailable {
                    AIModelRequiredBanner()
                }

                CodePatternView(code: """
                let guardrails: [any Guardrail] = [
                    InputLengthGuardrail(maxCharacters: \(maxCharacters), truncate: false),
                    ContentFilterGuardrail(blockedPatterns: \(blockedPatterns))
                ]
                let guarded = GuardedChain(
                    chain: SummarizeChain(),
                    guardrails: guardrails
                )
                let result = try await guarded.invoke(ChainInput(text: input))
                """)

                VStack(alignment: .leading, spacing: 8) {
                    Text("Input Text")
                        .font(.headline)
                    TextEditor(text: $inputText)
                        .frame(minHeight: 120)
                        .padding(8)
                        .background(Color.gray.opacity(0.15))
                        .clipShape(RoundedRectangle(cornerRadius: 8))

                    HStack {
                        Text("\(inputText.count) characters")
                            .font(.caption)
                            .foregroundStyle(inputText.count > maxCharacters ? .red : .secondary)
                        Spacer()
                        Text("Max: \(maxCharacters)")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                }

                VStack(alignment: .leading, spacing: 8) {
                    Text("Max Characters")
                        .font(.headline)
                    Stepper("\(maxCharacters)", value: $maxCharacters, in: 50...2000, step: 50)
                }

                VStack(alignment: .leading, spacing: 8) {
                    Text("Blocked Patterns (comma-separated)")
                        .font(.headline)
                    TextField("e.g. password, SSN", text: $blockedPatternsText)
                        .textFieldStyle(.roundedBorder)
                }

                Button(action: executeGuardedChain) {
                    HStack(spacing: 8) {
                        if isLoading {
                            ProgressView()
                                .scaleEffect(0.8)
                                .tint(.white)
                        }
                        Text(isLoading ? "Processing..." : "Summarize with Guardrails")
                    }
                    .frame(maxWidth: .infinity)
                    .frame(height: 20)
                }
                .buttonStyle(.borderedProminent)
                .disabled(isLoading || inputText.isEmpty || !isAIAvailable)

                if let errorMessage {
                    Text(errorMessage)
                        .foregroundStyle(.red)
                        .font(.caption)
                }

                if let blockedReason {
                    VStack(alignment: .leading, spacing: 8) {
                        Label("Blocked by Guardrail", systemImage: "shield.slash")
                            .font(.headline)
                            .foregroundStyle(.red)
                        Text(blockedReason)
                            .padding(12)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(Color.red.opacity(0.1))
                            .clipShape(RoundedRectangle(cornerRadius: 8))
                    }
                }

                if let resultText {
                    VStack(alignment: .leading, spacing: 8) {
                        Label("Guardrails Passed — Summary", systemImage: "shield.checkered")
                            .font(.headline)
                            .foregroundStyle(.green)
                        Text(resultText)
                            .padding(12)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(Color.green.opacity(0.1))
                            .clipShape(RoundedRectangle(cornerRadius: 8))
                    }
                }
            }
            .padding()
        }
    }

    private var blockedPatterns: [String] {
        blockedPatternsText
            .split(separator: ",")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
    }

    private func executeGuardedChain() {
        isLoading = true
        errorMessage = nil
        resultText = nil
        blockedReason = nil

        Task {
            do {
                let guardrails: [any Guardrail] = [
                    InputLengthGuardrail(maxCharacters: maxCharacters, truncate: false),
                    ContentFilterGuardrail(blockedPatterns: blockedPatterns),
                ]
                let guarded = GuardedChain(
                    chain: SummarizeChain(),
                    guardrails: guardrails
                )
                let output = try await guarded.invoke(ChainInput(text: inputText))
                await MainActor.run { resultText = output.text }
            } catch let error as LocanaraError {
                if case .invalidInput(let reason) = error {
                    await MainActor.run { blockedReason = reason }
                } else {
                    await MainActor.run { errorMessage = error.localizedDescription }
                }
            } catch {
                await MainActor.run { errorMessage = error.localizedDescription }
            }
            await MainActor.run { isLoading = false }
        }
    }
}

File: packages/android/example/src/main/kotlin/com/locanara/example/components/pages/framework/GuardrailDemo.kt
package com.locanara.example.components.pages.framework

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Block
import androidx.compose.material.icons.filled.Security
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.locanara.composable.ContentFilterGuardrail
import com.locanara.composable.GuardedChain
import com.locanara.composable.InputLengthGuardrail
import com.locanara.builtin.SummarizeChain
import com.locanara.core.ChainInput
import kotlinx.coroutines.launch

/**
 * Demonstrates Guardrails — wrapping chains with input validation and content safety.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun GuardrailDemo(onNavigateBack: () -> Unit) {
    var inputText by remember { mutableStateOf("Apple Intelligence is a personal intelligence system that puts powerful generative models at the core of iPhone, iPad, and Mac.") }
    var maxCharacters by remember { mutableFloatStateOf(500f) }
    var blockedPatternsText by remember { mutableStateOf("password, SSN, credit card") }
    var resultText by remember { mutableStateOf<String?>(null) }
    var blockedReason by remember { mutableStateOf<String?>(null) }
    var isLoading by remember { mutableStateOf(false) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    val scope = rememberCoroutineScope()

    val blockedPatterns = blockedPatternsText
        .split(",")
        .map { it.trim() }
        .filter { it.isNotEmpty() }

    val codePattern = """
val guardrails = listOf(
    InputLengthGuardrail(
        maxCharacters = ${maxCharacters.toInt()}, truncate = false
    ),
    ContentFilterGuardrail(
        blockedPatterns = listOf(${blockedPatterns.joinToString { "\"$it\"" }})
    )
)
val guarded = GuardedChain(
    chain = SummarizeChain(model),
    guardrails = guardrails
)
val result = guarded.invoke(ChainInput(text = input))
    """.trimIndent()

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Guardrail") },
                navigationIcon = {
                    IconButton(onClick = onNavigateBack) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back")
                    }
                }
            )
        }
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .padding(horizontal = 16.dp)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            CodePatternCard(code = codePattern)

            OutlinedTextField(
                value = inputText,
                onValueChange = { inputText = it },
                label = { Text("Input Text") },
                modifier = Modifier.fillMaxWidth(),
                minLines = 4,
                enabled = !isLoading
            )

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                Text(
                    "${inputText.length} characters",
                    style = MaterialTheme.typography.bodySmall,
                    color = if (inputText.length > maxCharacters.toInt()) MaterialTheme.colorScheme.error else MaterialTheme.colorScheme.onSurfaceVariant
                )
                Text("Max: ${maxCharacters.toInt()}", style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
            }

            Column {
                Text("Max Characters: ${maxCharacters.toInt()}", style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.Medium)
                Slider(
                    value = maxCharacters,
                    onValueChange = { maxCharacters = it },
                    valueRange = 50f..2000f,
                    steps = 38
                )
            }

            OutlinedTextField(
                value = blockedPatternsText,
                onValueChange = { blockedPatternsText = it },
                label = { Text("Blocked Patterns (comma-separated)") },
                modifier = Modifier.fillMaxWidth(),
                singleLine = true
            )

            Button(
                onClick = {
                    isLoading = true
                    errorMessage = null
                    resultText = null
                    blockedReason = null
                    scope.launch {
                        try {
                            println("[GuardrailDemo] input: ${inputText.take(200)}, maxChars: ${maxCharacters.toInt()}, blocked: $blockedPatterns")
                            val guardrails = listOf(
                                InputLengthGuardrail(maxCharacters = maxCharacters.toInt(), truncate = false),
                                ContentFilterGuardrail(blockedPatterns = blockedPatterns)
                            )
                            val guarded = GuardedChain(
                                name = "GuardedSummarize",
                                chain = SummarizeChain(),
                                guardrails = guardrails
                            )
                            val output = guarded.invoke(ChainInput(text = inputText))
                            resultText = output.text
                            println("[GuardrailDemo] result: ${resultText?.take(200)}")
                        } catch (e: IllegalArgumentException) {
                            println("[GuardrailDemo] blocked: ${e.message}")
                            blockedReason = e.message
                        } catch (e: Exception) {
                            println("[GuardrailDemo] error: ${e.message}")
                            errorMessage = e.message ?: "Unknown error"
                        } finally {
                            isLoading = false
                        }
                    }
                },
                modifier = Modifier.fillMaxWidth(),
                enabled = !isLoading && inputText.isNotBlank()
            ) {
                if (isLoading) CircularProgressIndicator(modifier = Modifier.padding(end = 8.dp), strokeWidth = 2.dp)
                Text(if (isLoading) "Processing..." else "Summarize with Guardrails")
            }

            errorMessage?.let { Text(it, color = MaterialTheme.colorScheme.error, style = MaterialTheme.typography.bodySmall) }

            blockedReason?.let { reason ->
                Card(
                    colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.errorContainer),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Icon(Icons.Default.Block, contentDescription = null, tint = MaterialTheme.colorScheme.error)
                            Text("  Blocked by Guardrail", style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.Medium, color = MaterialTheme.colorScheme.error)
                        }
                        Spacer(modifier = Modifier.height(8.dp))
                        Text(reason, style = MaterialTheme.typography.bodyMedium)
                    }
                }
            }

            resultText?.let { result ->
                Card(
                    colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.secondaryContainer),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Icon(Icons.Default.Security, contentDescription = null, tint = MaterialTheme.colorScheme.primary)
                            Text("  Guardrails Passed — Summary", style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.Medium)
                        }
                        Spacer(modifier = Modifier.height(8.dp))
                        Text(result, style = MaterialTheme.typography.bodyMedium)
                    }
                }
            }

            Spacer(modifier = Modifier.height(16.dp))
        }
    }
}
